### ISSUE TITLE
Agent.fsm : DeleteNode failed - Failed to insert usage entry for 'service-names' - Delta will cause a negative count

### ISSUE THREAD

**Issue by Michael-Kempe** [hashicorp/consul#9433]
#### Overview of the Issue

I see the following error in each of our cluster-server log:
[WARN]  agent.fsm: DeleteNode failed: error="failed to insert usage entry for "service-names": delta will cause a negative count"

Any hints on how to resolve this? 

### Consul info for both Client and Server

Server : v1.9.1
Client: v1.9.1

###output from server 'consul info' command:
agent:
        check_monitors = 0
        check_ttls = 0
        checks = 0
        services = 0
build:
        prerelease =
        revision = ca5c3894
        version = 1.9.1
consul:
        acl = enabled
        bootstrap = false
        known_datacenters = 1
        leader = false
        leader_addr = xxx.xxx.xxx.xxx:8300
        server = true
raft:
        applied_index = 862563
        commit_index = 862563
        fsm_pending = 0
        last_contact = 7.8479ms
        last_log_index = 862563
        last_log_term = 222
        last_snapshot_index = 852214
        last_snapshot_term = 215
        latest_configuration = [{Suffrage:Voter ID:8be12461-eb7e-4b71-836a-97014bcf43f4 Address:170.15.1.192:8300} {Suffrage:Voter ID:76d7ab59-f695-b929-9bf3-f06dd3743105 Address:170.15.1.193:8300} {Suffrage:Voter ID:96c6479d-b4bb-31a5-8c6e-af0bd7cf1ad4 Address:170.15.1.194:8300}]
        latest_configuration_index = 0
        num_peers = 2
        protocol_version = 3
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Follower
        term = 222
runtime:
        arch = amd64
        cpu_count = 8
        goroutines = 82
        max_procs = 8
        os = linux
        version = go1.15.6
serf_lan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 66
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 89352
        members = 4
        query_queue = 0
        query_time = 1
serf_wan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 717
        members = 3
        query_queue = 0
        query_time = 1

### Operating system and Environment details

Ubuntu 20.04.1 LTS

**Comment by crhino** [comment]
@Michael-Kempe Thanks for reporting, I took a look at the code and found an issue that exhibits the same warning message, see https://github.com/hashicorp/consul/pull/9440 for more details.

In terms of resolving this, it depends on the exact situation of your deployment unfortunately. Are you trying to remove a node from the deployment? Manually deregistering the service instances before removing the node is a workaround for this issue, if that is possible for you.

**Comment by Michael-Kempe** [comment]
My deployment is quite basic, as I'm trying to learn about consul. 
Having a cluster of 3 Server an 2 Client Agents and only 1 Service - (builtin 'consul' service). Nothing else.... 
So my question is: what service instances should I remove?  There are no nodes to remove.....

**Comment by crhino** [comment]
Sorry for the late response, was on holiday break. What have you used the cluster to do? It would be good to get a better understanding how you used the cluster in order to figure out what caused the error.

**Comment by ryan-dyer-sp** [comment]
@crhino I am also experiencing this issue.  Let me know if this doesnt sound the same and I'll open a different ticket.  For us, we have upgraded the consul version and helm chart we were using from 1+yrs ago.  We have a consul cluster dedicated as the backend for vault.  Post upgrade the consul cluster and vault cluster both appear to operate normally.  The only issue seen as that the catalog has references to the vault and consul Nodes prior the upgrade.  I was able to deregister the consul nodes, but one of the vault(and only) nodes I am unable to deregister.  new vault nodes are not showing up I believe due to this "bad" node still being present.  I attempted to deregister the serviceID for the Node as well, but also get an error
```
> k exec -it consul-server-0 -- curl -sk https://127.0.0.1:8501/v1/catalog/deregister -XPUT -d '{"Node": "vault-5d5d464fc7-bmrfk", "ServiceID": "vault:vault.vault.svc.cluster.local:8200"}'

consul-server-0 consul     2021-01-11T14:24:53.252Z [WARN]  agent.fsm: DeleteNodeService failed: error="failed to insert usage entry for "service-names": delta will cause a negative count"
```

**Comment by crhino** [comment]
@ryan-dyer-sp That certainly sounds similar, thanks for reporting it!

One action I would recommend, in order to fix the situation right now, is to restart the your Consul processes. That will force Consul to re-compute the usage metrics.

**Comment by ryan-dyer-sp** [comment]
@crhino I have tried restarting the consul-server pods and I still get the same error when trying to deregister the ServiceID.  At startup there are also alot of the error messages for DeleteNode.  We're running docker image hashicorp/consul:1.9.1 if that matters.

**Comment by crhino** [comment]
Interesting, I will look into that some more and get back to you!

**Comment by crhino** [comment]
@ryan-dyer-sp First of all, thanks for taking the time to respond to this issue, I looked into it more and saw the same issue causes the usage metrics to not recompute properly.

In terms of fixing this issue, unfortunately the easiest solution right now is to register fake service(s) to artificially bump the usage count, and then deregister the nodes that you are trying to remove. A complete fix should be in the next patch release.

**Comment by ryan-dyer-sp** [comment]
Glad to hear it will be fixed with the next patch release.  As we don't use the consul for anything other than as a backing store for vault, I don't believe there's any reason we need to try and get this fixed manually at this time.  We'll just wait for the patch.

**Comment by crhino** [comment]
Closing, this is fixed by https://github.com/hashicorp/consul/pull/9440 and released in 1.9.2.
